Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -265,434 +237,255 @@ final class ConnectionPool {
             case .none:
                 break
             }
         }
 
-        private func makeConnection(on eventLoop: EventLoop) -> EventLoopFuture<Connection> {
-            self.activityPrecondition(expected: [.opened])
-            let handshakePromise = eventLoop.makePromise(of: Void.self)
-            let bootstrap = ClientBootstrap.makeHTTPClientBootstrapBase(group: eventLoop, host: self.key.host, port: self.key.port, configuration: self.configuration)
-            let address = HTTPClient.resolveAddress(host: self.key.host, port: self.key.port, proxy: self.configuration.proxy)
-
-            let channel: EventLoopFuture<Channel>
-            switch self.key.scheme {
-            case .http, .https:
-                channel = bootstrap.connect(host: address.host, port: address.port)
-            case .unix:
-                channel = bootstrap.connect(unixDomainSocketPath: self.key.unixPath)
-            }
-
-            return channel.flatMap { channel -> EventLoopFuture<ConnectionPool.Connection> in
-                channel.pipeline.addSSLHandlerIfNeeded(for: self.key, tlsConfiguration: self.configuration.tlsConfiguration, handshakePromise: handshakePromise)
-                return handshakePromise.futureResult.flatMap {
-                    channel.pipeline.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes)
-                }.map {
-                    let connection = Connection(key: self.key, channel: channel, parentPool: self.parentPool)
-                    connection.isLeased = true
-                    return connection
-                }
-            }.map { connection in
-                self.configureCloseCallback(of: connection)
-                return connection
-            }.flatMapError { error in
-                // This promise may not have been completed if we reach this
-                // so we fail it to avoid any leak
-                handshakePromise.fail(error)
-                let action = self.parentPool.connectionProvidersLock.withLock {
-                    self.stateLock.withLock {
-                        self.state.failedConnectionAction()
-                    }
-                }
-                switch action {
-                case .makeConnectionAndComplete(let el, let promise):
-                    self.makeConnection(on: el).cascade(to: promise)
-                case .none:
-                    break
+        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
+            let waiter = Waiter(promise: self.eventLoop.makePromise(), preference: preference)
+
+            let action: Action = self.lock.withLock {
+                if let connection = self.availableConnections.popFirst() {
+                    return .lease(connection, waiter)
+                } else if self.openedConnectionsCount < self.maximumConcurrentConnections {
+                    self.openedConnectionsCount += 1
```

Comment:
This seems racy. We take a decision under the lock but then return without changing the state, so before we actually do the action, the circumstances may change. Or is this race okay?

Reply:
what do you mean by not changing state? we do increase the counter if we will be creating another connection. We only skip state change if we reuse connection

Reply:
Actually, we do change:
1. if we reuse - we pop from available connections,
2. if we create - we increase the connection count
3. if we cannot reuse or create - we append to waiters

Reply:
Sorry @artemredkin , not sure what I was thinking (or I can't see the problem I meant anymore) :)